astropy/io/misc/ecsv.py

```diff
@@ -0,0 +1,1245 @@
+    -------
+    data_out : numpy.ndarray or numpy.ma.MaskedArray
+        An array of objects reconstructed from `str_vals`, with the same shape as `col`.
+        If `mask` is provided, a masked array is returned with the mask applied.
+    col_shape : tuple[int, ...]
+        Expected shape of data_out, used in final sanity check of reading.
+    """
+    # Empty (blank) values in original ECSV are masked. Instead set the values
+    # to "[]" indicating an empty list. This operation also unmasks the values.
+    if mask is not None:
+        fill_value = "[]"
+        for idx in np.nonzero(mask)[0]:
+            str_vals[idx] = fill_value
+
+    # Remake as a 1-d object column of numpy ndarrays or
+    # MaskedArray using the datatype specified in the ECSV file.
+    col_vals = []
+    for str_val in str_vals:
+        obj_val = json.loads(str_val)  # list or nested lists
+        try:
+            arr_val = np.array(obj_val, dtype=col.dtype)
+        except TypeError:
+            # obj_val has entries that are inconsistent with
+            # dtype. For a valid ECSV file the only possibility
+            # is None values (indicating missing values).
+            vals = np.array(obj_val, dtype=object)
+            # Replace all the None with an appropriate fill value
+            mask_vals = vals == None
+            vals[mask_vals] = {"U": "", "S": b""}.get(col.dtype.kind, 0)
+            arr_val = np.ma.array(vals.astype(col.dtype), mask=mask_vals)
+
+        col_vals.append(arr_val)
+
+    np_empty = np.empty if mask is None else np.ma.empty
+    data_out = np_empty(len(col_vals), dtype=object)
+    data_out[:] = col_vals
+    if mask is not None:
+        data_out.mask = mask
+    return data_out, ()
+
+
+def get_null_values_per_column(
+    cols: list[ColumnECSV],
+    table_meta: dict | None,
+    null_values: list[str],
+) -> list[tuple[str, str, str]]:
+    """Get null and fill values for individual columns.
+
+    For ECSV handle the corner case of data that has been serialized using the
+    serialize_method='data_mask' option, which writes the full data and mask directly,
+    AND where that table includes a string column with zero-length string entries ("")
+    which are valid data. Normally the super() method will set col.fill_value=('', '0')
+    to replace blanks with a '0'.  But for that corner case subset, instead do not do
+    any filling.
+
+    Parameters
+    ----------
+    cols : list[ColumnECSV]
+        List of ColumnECSV objects representing the columns in the ECSV file.
+    table_meta : dict or None
+        Metadata dictionary from the ECSV header, which may include serialized columns
+        and other metadata.
+    null_values : list[str]
+        List of string values to interpret as null/missing values in the data. Default
+        is [""].
+
+    Returns
+    -------
+    fill_values : list[tuple[str, str, str]]
+        A list of tuples with (null_value, fill_value, column_name) for each column in
```

Comment:
This is not strictly true, I think: if there are serialized `MaskedColumn`, those are not included. I.e., `len(fill_values) <= len(cols)`

Actually, this does feel rather fragile. Isn't it better to just include an empty tuple in that case so that the lengths are guaranteed to match? Or return a `dict` keyed by the column name? (An advantage of the latter would be that one doesn't need to store `col.name` in the tuple any more.)